src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PivotUpdator.cs

```diff
@@ -0,0 +1,249 @@
+    private Keccak _alreadyAnnouncedNewPivotHash = Keccak.Zero;
+
+    public PivotUpdator(IBlockTree blockTree,
+        ISyncModeSelector syncModeSelector,
+        ISyncPeerPool syncPeerPool,
+        ISyncConfig syncConfig,
+        IBlockCacheService blockCacheService,
+        IBeaconSyncStrategy beaconSyncStrategy,
+        IDb metadataDb,
+        ILogManager logManager)
+    {
+        _blockTree = blockTree ?? throw new ArgumentNullException(nameof(blockTree));
+        _syncModeSelector = syncModeSelector ?? throw new ArgumentNullException(nameof(syncModeSelector));
+        _syncPeerPool = syncPeerPool ?? throw new ArgumentNullException(nameof(syncPeerPool));
+        _syncConfig = syncConfig ?? throw new ArgumentNullException(nameof(syncConfig));
+        _blockCacheService = blockCacheService ?? throw new ArgumentNullException(nameof(blockCacheService));
+        _beaconSyncStrategy = beaconSyncStrategy ?? throw new ArgumentNullException(nameof(beaconSyncStrategy));
+        _metadataDb = metadataDb ?? throw new ArgumentNullException(nameof(metadataDb));
+        _logger = logManager.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+
+        _attemptsLeft = syncConfig.MaxAttemptsToUpdatePivot;
+
+        if (!TryUpdateSyncConfigUsingDataFromDb())
+        {
+            _syncModeSelector.Changed += OnSyncModeChanged;
+        }
+    }
+
+    private bool TryUpdateSyncConfigUsingDataFromDb()
+    {
+        try
+        {
+            if (_metadataDb.KeyExists(MetadataDbKeys.UpdatedPivotNumber) && _metadataDb.KeyExists(MetadataDbKeys.UpdatedPivotHash))
+            {
+                byte[]? numberFromDb = _metadataDb.Get(MetadataDbKeys.UpdatedPivotNumber);
+                RlpStream numberStream = new(numberFromDb!);
+                long updatedPivotBlockNumber = numberStream.DecodeLong();
+
+                byte[]? hashFromDb = _metadataDb.Get(MetadataDbKeys.UpdatedPivotHash);
+                RlpStream hashStream = new(hashFromDb!);
+                Keccak updatedPivotBlockHash = hashStream.DecodeKeccak()!;
+
+                _syncConfig.PivotNumber = updatedPivotBlockNumber.ToString();
+                _syncConfig.PivotHash = updatedPivotBlockHash.ToString();
+                _syncConfig.MaxAttemptsToUpdatePivot = 0;
+
+                if (_logger.IsWarn) _logger.Warn($"Pivot block has been set based on data from db. Pivot block number: {updatedPivotBlockNumber}, hash: {updatedPivotBlockHash}");
+                return true;
+            }
+        }
+        catch (RlpException)
+        {
+            if (_logger.IsWarn) _logger.Warn($"Cannot decode pivot block number or hash");
+        }
+
+        return false;
+    }
+
+    private async void OnSyncModeChanged(object? sender, SyncModeChangedEventArgs syncMode)
+    {
+        if ((syncMode.Current & SyncMode.UpdatingPivot) != 0 && Interlocked.Read(ref _updateInProgress) == 0)
+        {
+            Interlocked.Increment(ref _updateInProgress);
```

Comment:
Interlocked.Read and later Interlock.Increment are not thread safe. You probably want Interlock.CompareExchenge.

Reply:
Thanks for catching it, fixed!